src/tree/updater_quantile_hist.h

```diff
@@ -28,6 +28,38 @@
 #include "../common/column_matrix.h"
 
 namespace xgboost {
+
+template<typename T, size_t MaxStackSize>
+class MemStackAllocator {
```

Comment:
std::vector - very not efficient, but probably in this case it will work.
if you come back tbb - they have std-style ScalableAllocator which can be useful here.